In [2]:
import numpy as np
import pandas as pd 
from datetime import datetime

In [22]:
df = pd.read_csv("../data/mtd_combined.csv", index_col=0)
df.head()
df.to_excel("test.xlsx")

In [4]:
df_stops = pd.read_excel("../data/MoDstops+Preismodell.xlsx", sheet_name="MoDstops")
df_stops.head()

MoDStop Id             MoDStop Name  MoDStop Lat  MoDStop Long  \
0        1001              Mandelgasse    49.351780      8.129000   
1        1002          Hauptfeuerwache    49.353733      8.131552   
2        1003  Kindergarten St. Marien    49.355745      8.134341   
3        1004               Strohmarkt    49.355520      8.136674   
4        1005                  Rathaus    49.354022      8.135143   

                   MoDStop Adresse  
0     Seilerbahn 1, 67433 Neustadt  
1  Lindenstraße 11, 67433 Neustadt  
2  Sauterstraße 12, 67433 Neustadt  
3       Strohmarkt, 67433 Neustadt  
4  Rathausstraße 1, 67433 Neustadt

### Column: id

In [5]:
print("Total number of ID's:", len(df.id))
print("# of NaN:", int(df[['id']].isna().sum()))
print("# of duplicates:", df.id.duplicated().sum() - int(df[['id']].isna().sum() - 1), "(exklusive NaN)")

Total number of ID's: 10348
# of NaN: 4535
# of duplicates: 0 (exklusive NaN)


4535 fehlende ID's, alle mit dem Status offer oder offer_rejeceted. Falls wir diese als Primärschlüssel nutzen wollen evtl. künstlich hinzufügen?

### Column: user_id

In [6]:
print("# of NaN:", int(df[['user_id']].isna().sum()))
print("# of duplicates:", df.user_id.duplicated().sum() - int(df[['user_id']].isna().sum()), "(exklusive NaN)")

# of NaN: 0
# of duplicates: 9565 (exklusive NaN)


### Column: distance

In [7]:
print("# of NaN:", int(df[['distance']].isna().sum()))
print("# of neg:", len(df.loc[(df.distance < 0)]))
print("# of zero:", len(df.loc[(df.distance == 0)]))
print("# of ones:", len(df.loc[(df.distance == 1)]))

# of NaN: 0
# of neg: 0
# of zero: 0
# of ones: 14


In [8]:
comparison = np.where(df['pickup_address'] == df['dropoff_address'], True, False)
df.loc[:, 'isSameAddress'] = comparison
mask = df['distance']==1
df[mask][['pickup_address', 'dropoff_address', 'isSameAddress']]

pickup_address       dropoff_address  isSameAddress
37        49.322557|8.12336     49.322557|8.12336           True
49          Haardter Winzer       Haardter Winzer           True
114          Am Herzel West        Am Herzel West           True
418             TC Rot-Weiß           TC Rot-Weiß           True
9502                 Schütt                Schütt           True
9563     49.330245|8.130933    49.330245|8.130933           True
9697     49.351731|8.136637    49.351731|8.136637           True
9828     49.376347|8.153444    49.376347|8.153444           True
9903     49.316193|8.135477    49.316193|8.135477           True
10020                Schütt                Schütt           True
10227     49.35552|8.136674     49.35552|8.136674           True
10294    49.376347|8.153444    49.376347|8.153444           True
10297  Le Quartier Hornbach  Le Quartier Hornbach           True
10331            Strohmarkt            Strohmarkt           True

### Column: number_of_passenger

In [9]:
print("# of NaN:", int(df[['number_of_passenger']].isna().sum()))
print("# of neg:", len(df.loc[(df.number_of_passenger < 0)]))
print("# of zero:", len(df.loc[(df.number_of_passenger == 0)]))
print("# of pos:", len(df.loc[(df.number_of_passenger > 0)]))
print('Max # of passengers:', df['number_of_passenger'].max())

# of NaN: 4535
# of neg: 0
# of zero: 0
# of pos: 5813
Max # of passengers: 5.0


NaN Zeilen sind auf die offer oder offer_rejected zurückzuführen. Wenn man diese beachtet könnte man 0 hinzufügen.

### Column: price_operations

In [10]:
print("# of NaN:", int(df[['price_operations']].isna().sum()))
print("# of neg:", len(df.loc[(df.price_operations < 0)]))
print("# of zero:", len(df.loc[(df.price_operations == 0)]))
print("# of pos:", len(df.loc[(df.price_operations > 0)]))
print('Min price:', df['price_operations'].min())
print('Max price:', df['price_operations'].max())

# of NaN: 1204
# of neg: 0
# of zero: 458
# of pos: 8686
Min price: 0.0
Max price: 33.1


### Column: price_offer

In [11]:
print("# of NaN:", int(df[['price_offer']].isna().sum()))
print("# of neg:", len(df.loc[(df.price_offer < 0)]))
print("# of zero:", len(df.loc[(df.price_offer == 0)]))
print("# of pos:", len(df.loc[(df.price_offer > 0)]))
print('Min price:', df['price_offer'].min())
print('Max price:', df['price_offer'].max())

# of NaN: 4535
# of neg: 0
# of zero: 0
# of pos: 5813
Min price: 0.8
Max price: 32.3


### Column: price_payed

In [12]:
print("# of NaN:", int(df[['price_payed']].isna().sum()))
print("# of neg:", len(df.loc[(df.price_payed < 0)]))
print("# of zero:", len(df.loc[(df.price_payed == 0)]))
print("# of pos:", len(df.loc[(df.price_payed > 0)]))
print('Min price:', df['price_payed'].min())
print('Max price:', df['price_payed'].max())

# of NaN: 4535
# of neg: 0
# of zero: 1322
# of pos: 4491
Min price: 0.0
Max price: 555.33


### Column: free_ride

In [17]:
print("# of NaN:", int(df[['free_ride']].isna().sum()))
df.free_ride.value_counts()

# of NaN: 4535


0.0    5170
1.0     643
Name: free_ride, dtype: int64

### Column: payment_type

### Column: bahn_card_number

### Column: year_card_type

### Column: year_card_number

### Column: pickup_address and dropoff_address

In [18]:
print("# of NaN:", int(df[['pickup_address']].isna().sum()))
print("# of NaN:", int(df[['dropoff_address']].isna().sum()))

# of NaN: 0
# of NaN: 0


In [19]:
df['start_id'] = ""
df['end_id'] = ""

def get_stop_id(address):
    if address[0].isdigit():
        lat = address.split("|")[0]
        long = address.split("|")[1]
        for index, row in df_stops.iterrows():
            if str(row["MoDStop Lat"]) == lat and str(row["MoDStop Long"]) == long:
                return row["MoDStop Id"]
        return "No match of lat and long"
    else:
        # different naming
        if address == "Rewe Mußbach":
            address = address + " (Shoppenwiese)"
        for index, row in df_stops.iterrows():
            if row["MoDStop Name"] == address:
                return row["MoDStop Id"]
        return "No match of address name"

In [20]:
for index, row in df.iterrows():
    df.at[index, 'start_id'] = get_stop_id(row['pickup_address'])
    df.at[index, 'end_id'] = get_stop_id(row['dropoff_address'])

### Column: state

### Column: created_from_offer

### Column: created_at

### Column: scheduled_to

### Column: dispatched_at

### Column: canceled_at

### Column: pickup_arrival_time

### Column: arriving_push

### Column: vehicle_arrived_at

### Column: earliest_pickup_expectation

### Column: pickup_first_eta

### Column: pickup_at

### Column: dropoff_first_eta

### Column: dropoff_eta

### Column: dropoff_at

### Column: updated_at

### Column: arrival_destination

### Column: waiting_time

### Column: boarding_time

### Column: ride_time

### Column: trip_time

### Column: shortest_ridetime

### Column: delay

### Column: longer_route_factor

### Column: arrival indicator